vllm/model_executor/parallel_utils/tensor_parallel/layers.py

```diff
@@ -427,17 +433,12 @@ def forward(self, input_):
             input_parallel = input_
         else:
             input_parallel = scatter_to_tensor_model_parallel_region(input_)
-        if get_tensor_model_parallel_world_size() == 1:
-            # Matrix multiply.
-            output_ = F.linear(input_parallel, self.weight)
+        # Matrix multiply.
+        output_parallel = F.linear(input_parallel, self.weight)
+        if self.reduce_results:
+            output_ = reduce_from_tensor_model_parallel_region(output_parallel)
```

Comment:
To my understanding, this can incur unnecessary overhead when `get_tensor_model_parallel_world_size() == 1`, similarly to what we found in #146. In general, we should avoid calling any `autograd.Function` when it is actually a no-op.

Reply:
Fixed!